### ISSUE TITLE
Incorrect parsing of sigprocmask (multiple expressions)

### ISSUE THREAD

**Issue by ZetaTwo** [desbma/shh#3]
This line will not be parsed correctly:
```
"440663      0.002174 rt_sigprocmask(SIG_SETMASK, [], ~[KILL STOP RTMIN RT_1], 8) = 0"
```

The problem is that the "array" group will match all the way from the first "[" to the last "]", see https://regex101.com/r/xbQbek/1
I tried coming up with an easy fix but I'm worried that you can have expressions inside arrays and that would cause problems. I don't know strace well enough to know if this is an issue. If it is not, then maybe changing
```
(
        \[
        (?<array>.+)
        \]
    )
```

to 
```
(
        \[
        (?<array>[~\]]+)
        \]
    )
```

could work.

**Comment by ZetaTwo** [comment]
I started writing a [Pest]() grammar to try to see if it would make sense to replace the parser. It's still not complete but it's a start:
```
EQ     = _{ "=" }
SP     = _{ " " }
PERIOD = _{ "." }
NOT_LF = _{ !NEWLINE ~ ANY }

DIGITS        = _{ ASCII_DIGIT+ }
HEX_NUMBER = _{ "0x" ~ ASCII_HEX_DIGIT+ }
DEC_NUMBER = _{ ASCII_NONZERO_DIGIT ~ ASCII_DIGIT* }
TIMESTAMP     = _{ DIGITS ~ PERIOD ~ DIGITS }
DURATION      = _{ DIGITS ~ PERIOD ~ DIGITS }
ALPHANUMUNDER = _{ ASCII_ALPHANUMERIC | "_" }
ALPHAUNDER    = _{ ASCII_ALPHA | "_" }

start = { (NEWLINE | line)+ }

line      = { pid ~ SP+ ~ timestamp ~ SP ~ body ~ NEWLINE }
pid       = { DIGITS }
timestamp = { TIMESTAMP }
body      = { syscall | alert_body }

syscall      = { syscall_name ~ "(" ~ syscall_args ~ ")" ~ SP+ ~ "=" ~ SP+ ~ syscall_result ~ (SP ~ syscall_duration)? }
syscall_name = { ALPHANUMUNDER+ }
// -> name
syscall_args = { syscall_arg? ~ ("," ~ SP ~ syscall_arg)* }
// -> args
syscall_arg = {
  
  | sigset
  | braced
  | bracketed
  | expression_or
  | expanded_ts
  | number
  | string
  | named_constant
  | key_value
  | function_like
  | other
}

syscall_result = { NOT_LF+ }
// -> result
syscall_duration = { "<" ~ DURATION ~ ">" }
// -> duration

datetime = { ASCII_DIGIT{4} ~ "-" ~ ASCII_DIGIT{2} ~ "-" ~ ASCII_DIGIT{2} ~ "T" ~ ASCII_DIGIT{2} ~ ":" ~ ASCII_DIGIT{2} ~ ":" ~ ASCII_DIGIT{2} ~ "." ~ ASCII_DIGIT+ ~ ("+" ~ ASCII_DIGIT{4})? } 
string = { "\"" ~ (!"\"" ~ ANY)* ~ "\"" }
expanded_ts = { number ~ SP ~ "/*" ~ SP ~ datetime ~ SP ~ "*/" }


number         = { DEC_NUMBER | HEX_NUMBER | DIGITS }
named_constant = { (DIGITS | UPPERCASE | "_")+ }
bracketed      = { "[" ~ syscall_args ~ "]" }

expression_or = { (number | named_constant) ~ ("|" ~ (number | named_constant))+ }
braced         = { "{" ~ syscall_args ~ "}" }
key_value      = { key_value_key ~ "=" ~ key_value_value }
key_value_key  = { ALPHAUNDER+ }
// -> key
key_value_value = { syscall_arg }
// -> value
function_like = { ALPHAUNDER+ ~ "(" ~ syscall_args ~ ")" }

NEGATED = { "~" }
SIGNAL  = { (ASCII_DIGIT | ASCII_ALPHA_UPPER | "_")+ }
sigset  = { NEGATED? ~ "[" ~ SIGNAL? ~ (SP ~ SIGNAL)* ~ "]" }
// Anything else
other = { !("," | ")" | "}" | "]")+ }

alert_body = { "+++" ~ SP ~ "exited with" ~ SP ~ DIGITS ~ SP ~ "+++" }
```

**Comment by desbma** [comment]
Thanks for the report, your interpretation of the problem and the suggested fix is correct. 

It should be fixed in be5dd32.

Please open a new issue for more general discussion about the strace parser.
